### PR TITLE
Refactor leader election

### PR DESCRIPTION
**Pull request by chadpatel**
**Description:** Refactor leader election out of k8sapiserver.go so that it can be utilized by multiple consumers of the k8sapiserver APIs (i.e. prometheus scraper)

I also updated the receiver and scraper(s) to use the new mechanism.  Unit tests were shuffled around and new tests were added.

Builds on this (still unmerged) PR
https://github.com/amazon-contributing/opentelemetry-collector-contrib/pull/16/commits

**Link to tracking Issue:** N/A

**Testing:** 
I verified the metrics are getting uploaded from a single leader.  Example:
Screen Shot 2023-05-30 at 3 51 49 PM

Metrics are flowing:
Screen Shot 2023-05-30 at 3 54 26 PM

**Documentation:** Updated

### CODE REVIEW

File: receiver/awscontainerinsightreceiver/internal/k8sapiserver/leaderelection_test.go [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+	mockClient.On("ClusterFailedNodeCount").Return(1)
+	mockClient.On("ClusterNodeCount").Return(1)
+	mockClient.On("ServiceToPodNum").Return(
+		map[k8sclient.Service]int{
+			NewService("service1", "kube-system"): 1,
+			NewService("service2", "kube-system"): 1,
+		},
+	)
+
+	<-leaderElection.isLeadingC
```
Comment (SaxyPandaBear) [inline receiver/awscontainerinsightreceiver/internal/k8sapiserver/leaderelection_test.go:82]: nit - This blocks indefinitely. Typically when I do this with channels in tests, I wrap it in a select with a timeout just to make sure the test doesn't hang.
```go
select {
    case b := <-leaderElection.isLeadingC:
        assert.True(t, b)
    case time.After(100 * time.Millisecond):
        t.Errorf("failed to elect leader after 100ms")
}
```

File: receiver/awscontainerinsightreceiver/internal/k8sapiserver/prometheus_scraper.go [hunk 2]
```diff
@@ -42,23 +42,11 @@ type PrometheusScraper struct {
...
-	}
-
-	// get endpoint
-	endpoint := k8sClient.GetClientSet().CoreV1().RESTClient().Get().AbsPath("/").URL().Hostname()
-
-	// TODO: test RBAC permissions?
-
-	spFactory := simpleprometheusreceiver.NewFactory() // TODO: pass this in?
-
+func NewPrometheusScraper(ctx context.Context, telemetrySettings component.TelemetrySettings, endpoint string, nextConsumer consumer.Metrics, host component.Host, clusterNameProvider clusterNameProvider, leaderElection *LeaderElection) (*PrometheusScraper, error) {
```
Comment (SaxyPandaBear) [inline receiver/awscontainerinsightreceiver/internal/k8sapiserver/prometheus_scraper.go:49]: nit should we validate that the leaderElection is not nil?

**Review by SaxyPandaBear — APPROVED** [review]

File: receiver/awscontainerinsightreceiver/internal/k8sapiserver/leaderelection_test.go (hunk 1 above)
Comment (chadpatel) [reply receiver/awscontainerinsightreceiver/internal/k8sapiserver/leaderelection_test.go:82]: Love this! I will do this in my follow up where I introduce the config change

File: receiver/awscontainerinsightreceiver/internal/k8sapiserver/prometheus_scraper.go (hunk 2 above)
Comment (chadpatel) [reply receiver/awscontainerinsightreceiver/internal/k8sapiserver/prometheus_scraper.go:49]: Good call, I am adding that behavior and a test in my follow up PR

**Review by sky333999 — APPROVED** [review]

File: receiver/awscontainerinsightreceiver/internal/k8sapiserver/leaderelection.go [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+	StartRecordingToSink(sink record.EventSink) watch.Interface
+	// StartLogging starts sending events received from this EventBroadcaster to the given logging
+	// function. The return value can be ignored or used to stop recording, if desired.
+	StartLogging(logf func(format string, args ...interface{})) watch.Interface
+	// NewRecorder returns an EventRecorder that can be used to send events to this EventBroadcaster
+	// with the event source set to the given event source.
+	NewRecorder(scheme *runtime.Scheme, source v1.EventSource) record.EventRecorder
+}
+
+type K8sClient interface {
```
[the remark this answers is not in the document]
Comment (chadpatel) [reply receiver/awscontainerinsightreceiver/internal/k8sapiserver/leaderelection.go:76]: Yeah, maintaining that specific structure ended up being easier.  Refactoring that would also be a decent sized change to the tests.  I agree it is awkward

File: receiver/awscontainerinsightreceiver/internal/k8sapiserver/prometheus_scraper.go [hunk 4]
```diff
@@ -92,19 +81,34 @@ func NewPrometheusScraper(ctx context.Context, telemetrySettings component.Telem
...
 func (ps *PrometheusScraper) Shutdown() {
-	err := ps.simplePrometheusReceiver.Shutdown(ps.ctx)
-	if err != nil {
-		panic("fix me when productionalizing code")
+	if ps.running {
+		err := ps.simplePrometheusReceiver.Shutdown(ps.ctx)
+		if err != nil {
+			ps.settings.Logger.Error("Unable to shutdown SimplePrometheusReceiver", zap.Error(err))
+		}
+		ps.running = err != nil
```
[the remark this answers is not in the document]
Comment (chadpatel) [reply receiver/awscontainerinsightreceiver/internal/k8sapiserver/prometheus_scraper.go:112]: Yeah, that is weird.  I will change it to this in my follow up PR
```
		ps.running = false
```

**Pull request merged**